### PR TITLE
Test virtual device & test firmware

### PR DESCRIPTION
**Pull request by francois141**

### CODE REVIEW

File: src/host.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ pub struct MiralisContext {
     /// Hardware capabilities of the core (hart).
     pub hw: HardwareCapability,
     /// List of devices with PMP
-    pub devices: [device::VirtDevice; 1],
+    pub devices: [device::VirtDevice; 2],
```
Comment (francois141) [inline src/host.rs:19]: @CharlyCst Do we want to have a test device on each platform?

Comment (CharlyCst) [reply src/host.rs:19]: Let's say yes for now, I don't expect it to add much overhead in terms of space or performance.
In the future we will need to have a better way of describing platforms and their devices, but we don't have enough devices to worry about it yet :)

File: src/device/clint.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
+use spin::Mutex;
+
+use crate::device::{DeviceAccess, Width};
+use crate::driver::clint::{
+    MSIP_OFFSET, MSIP_WIDTH, MTIMECMP_OFFSET, MTIMECMP_WIDTH, MTIME_OFFSET,
+};
+use crate::driver::ClintDriver;
+// ————————————————————————————— Virtual CLINT —————————————————————————————— //
```
Comment (CharlyCst) [inline src/device/clint.rs:8]: Style nit: we usually add a space before comment separators ^^
```suggestion

// ————————————————————————————— Virtual CLINT —————————————————————————————— //
```

File: src/device/mod.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
+//! Base device classes
+
+pub mod clint;
+pub mod tester;
+// ———————————————————————————— Virtual Devices ————————————————————————————— //
```
Comment (CharlyCst) [inline src/device/mod.rs:5]: Same style remark, we add a space before here
```suggestion

// ———————————————————————————— Virtual Devices ————————————————————————————— //
```

File: src/device/tester.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+use core::sync::atomic::{AtomicUsize, Ordering};
+
+use crate::device::{DeviceAccess, Width};
+
+// ————————————————————————————— Virtual Test Device —————————————————————————————— //
+
+pub const TEST_DEVICE_SIZE: usize = 0x8;
+
+#[derive(Debug)]
+pub struct VirtTestDevice {
```
Comment (CharlyCst) [inline src/device/tester.rs:10]: Maybe we can have a small doc comment here to explain what the purpose of this device is (testing).

**Review by CharlyCst — COMMENTED** [review]
The code looks good to me :)

Does the test pass? If yes we can add it to `just test` to run it in the CI

**Comment by francois141** [comment]
@CharlyCst Do you know what could go wrong here....?

Capture d’écran 2024-08-29 à 15 53 18

**Comment by CharlyCst** [comment]
Wow interesting, both numbers are printed as being exactly the same?

Assuming the device emulation indeed returns the right value, it is possible that somehow we are invoking undefined behavior here, maybe by polluting some registers... I would try to see if the state is corrupted when coming back from device emulation.
One way to do that would be to look at the assemble and step through it with GDB (one breakpoint before the read, one breakpoint after and see if all registers make sense). Another option is to try to rewrite the assert as an `if` branch and a panic and see if it still crashed, if that is the case looking at the assembly of the `if` should be simpler than the assert and GDB can probably give an answer.

**Comment by CharlyCst** [comment]
Hmmm, something strange is going on here, I am not sure but it might be related to #154.
I'd like to investigate a bit before merging this, I'll take a look and see if I find anything suspicious.

**Comment by CharlyCst** [comment]
The issue seems to be a bug in our load emulation (`lw` in particular).
The spec says that `lw` should perform a sign extension, which is not working properly in our emulation. As a result the test tries to compare `0xffffffffdeadbeef` (loaded from the `0xdeadbeef` hardcoded constant) and `0xdeadbeef` from reading the test device. The write should sign extend to `0xffffffffdeadbeef` so that the comparison can be done properly.

I'll investigate the root cause a bit further.

**Comment by CharlyCst** [comment]
So turns out the test driver caught a bug in our load emulation ^^
I pushed a fix, the test passes properly with `0xdeadbeef` now.

**Pull request merged**